pkg/apis/resource/types.go

```diff
@@ -17,17 +17,267 @@ limitations under the License.
+	Finalizer = "resource.kubernetes.io/delete-protection"
+)
+
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// ResourceClaim describes which resources are needed by a resource consumer.
-// Its status tracks whether the resource has been allocated and what the
-// resulting attributes are.
+// ResourceSlice represents one or more resources in a pool of similar resources,
+// managed by a given driver. A pool may span more than one ResourceSlice, and exactly how many
+// ResourceSlices comprise a pool is determined by the driver.
+//
+// At the moment, the only supported resources are devices with attributes and capacities.
+// Each device in a given pool, regardless of how many ResourceSlices, must have a unique name.
+// The ResourceSlice in which a device gets published may change over time. The unique identifier
+// for a device is the tuple <driver name>, <pool name>, <device name>.
+//
+// Whenever a driver needs to update a pool, it increments the pool.Spec.Pool.Generation number
+// and updates all ResourceSlices with that new number and new resource definitions. A consumer
+// must only use ResourceSlices with the highest generation number and ignore all others.
+//
+// When allocating all resources in a pool matching certain criteria or when
+// looking for the best solution among several different alternatives, a
+// consumer should check the number of ResourceSlices in a pool (included in
+// each ResourceSlice) to determine whether its view of a pool is complete and
+// if not, should wait until the driver has completed updating the pool.
+//
+// For resources that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceSlice struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec automatically increments the metadata.generation number.
+	Spec ResourceSliceSpec
+}
+
+const (
+	// ResourceSliceSelectorNodeName can be used in a [metav1.ListOptions]
+	// field selector to filter based on [ResourceSliceSpec.NodeName].
+	ResourceSliceSelectorNodeName = "spec.nodeName"
+	// ResourceSliceSelectorDriver can be used in a [metav1.ListOptions]
+	// field selector to filter based on [ResourceSliceSpec.Driver].
+	ResourceSliceSelectorDriver = "spec.driver"
+)
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// Driver identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	Driver string
```

Comment:
Perhaps it's just me, but I see "the kubelet plugin" of a DRA driver as only one of its pieces. Even without classic DRA and its control plane controller it might still have pieces for validation of configs (either configuration via VAP or some actual webhook).

As you said, "kubelet plugin" also isn't a good name.

Reply:
I'd vote for using "Kubelet DRA plugin" as in my understanding it is a Kubelet plugin as it uses Kubelet plugin registration APIs also used by Device and CSI **plugins**. I agree with @johnbelamaric that "driver" sounds much more confusing.

Reply:
We went back and forth on this many times in various objects when working on "classic" DRA. I know we had `KubeletPlugin` in one of the objects at one point and @thockin suggested we change it to `Driver`, so we did. I personally, don't feel too strongly about the name, but we need to pick one and commit to it.

Reply:
CSI doesn't use the word "plugin". It uses "driver", just like DRA does:

https://kubernetes-csi.github.io/docs/drivers.html

Reply:
So CSI is actually an argument **in favor** of keeping the "driver" terminology, not **against** it...

> it is a Kubelet plugin as it uses Kubelet plugin registration

That is an implementation detail that most users will not be aware of.

Reply:
In https://github.com/kubernetes/kubernetes/pull/125488/files#discussion_r1670044997 @klueska pointed out that for the control plane controller we are more specific and call it "controller" because that is the function that gets referenced.

If this field here was part of the allocation result, then *maybe* I could be convinced to call the field "plugin" there in analogy to the "controller" field because that's another function of the driver.

But this here is not necessarily a kubelet plugin. If the driver is managing network-attached devices, then the "the DRA driver providing the capacity information" is not a kubelet plugin. Calling it "plugin" would be wrong, unless we interpret the word as "plugin for the cluster", which is just confusing.

Therefore I am against renaming this field here. When focusing only on structured parameters, we don't have the "controller" field. I therefore prefer to keep the structured parameters API consistent and use "driver" everywhere, with "controller" the only exception for classic DRA.

Reply:
Makes sense to me, let's stick with driver.